Show "Collapse All" command in tree view toolbar.

tsmaeder opened the pull request:
Contributed on behalf of STMicroelectronics

<!--
Thank you for your Pull Request. Please provide a description and review
the requirements below.

Contributors guide: https://github.com/theia-ide/theia/blob/master/CONTRIBUTING.md

Note: Security vulnerabilities should not be disclosed on GitHub, through a PR or any
other means. See SECURITY.md at the root of this repository, to learn how to report
vulnerabilities.
-->

#### What it does

Closes https://github.com/eclipse-theia/theia/issues/12513.

Adds a "Collapse All" toolbar item in tree views contributed by plugins if the plugin requests it in the options (`showCollapseAll`). This PR introduces the concept of `DynamicWidget`. It allows to update the toolbar after the contents of a view widget has been create lazily.

#### How to test
Use the procedure from the related issue and make sure the icon appears. You can also use the built-in "NPM SCRIPTS" view as a second test case.

#### Review checklist

- [x] As an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- As a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

martin-fleck-at commented on packages/plugin-ext/src/main/browser/plugin-frontend-contribution.ts:
```diff
@@ -15,21 +15,56 @@
+        commands.registerCommand(PluginApiFrontendContribution.COLLAPSE_ALL_COMMAND, {
+            execute: (widget: Widget) => {
+                if (widget instanceof PluginViewWidget && widget.widgets[0] instanceof TreeViewWidget) {
+                    const model = widget.widgets[0].model;
```
Do you think it might be valuable to have that 'collapseTree' functionality generically in the model or is it specific to the tree that is usually provided through the plugins? To me, it looks quite generic but I know that roots might be handled differently depending on the expected tree structure.

tsmaeder replied:
Yes, this sounds quite generic, but I don't see that refactoring as a necessary part of this PR.

martin-fleck-at replied:
Understood.

martin-fleck-at commented on packages/plugin-ext/src/main/browser/plugin-frontend-contribution.ts:
```diff
@@ -15,21 +15,56 @@
 
+    static readonly COLLAPSE_ALL_COMMAND = Command.toDefaultLocalizedCommand({
+        id: 'treeviews.collapseAll',
+        iconClass: codicon('collapse-all')
```
```suggestion
        iconClass: codicon('collapse-all'),
        label: nls.localizeByDefault('Collapse All')
```

Otherwise we don't get any hover info on the toolbar button.

martin-fleck-at commented on packages/plugin-ext/src/main/browser/plugin-frontend-contribution.ts:
```diff
@@ -15,21 +15,56 @@
     }
 
+    registerToolbarItems(registry: TabBarToolbarRegistry): void {
+
```
nit: Unnecessary empty line?

martin-fleck-at commented on packages/core/src/browser/view-container.ts:
```diff
@@ -76,6 +76,20 @@ export namespace BadgeWidget {
+ * updating the toolbar when a contributed view is contructed "lazily"
+ */
+export interface DynamicWidget {
+    onChanged: CommonEvent<void>;
```
I really like the idea of having a dynamic widget that is able to trigger an update on it's own toolbar but I was wondering about some things:
- Isn't it conceptually very close to the `onDidChange` event on the `ConditionalToolbarItem`, is there any way to re-use that or is it because of the indirection through the ViewContainerPart?
- The name is quite generic even through it has a very specific use case (at the moment), do we really want to rely on such a generic name that might be inadvertently already implemented on a widget? Also I believe we should follow the `onDidX` naming scheme here, according to the [Coding Guidelines](https://github.com/eclipse-theia/theia/wiki/Coding-Guidelines).

tsmaeder replied:
The answer to both, IMO, is no. What's changing here is not a single toolbar item, but the set of toolbar items that apply for this view. The concrete case I'm covering is that the "real" widget that forms the context for action enablement is created lazily and we need to update the toolbar when that happens. However, that lazy creation is happening inside the view widget, and is an implementation detail of the view widget which I don't want to expose. 
I know the terminology is vague, but intentionally so. Once have more cases and understand the issue better, a better naming may become clear.

martin-fleck-at reviewed (changes requested):
Hi @tsmaeder,

Thank you very much for that change! I tested it with a single-root and a multi-root tree and both times the behavior was as expected. The flag is also correctly considered (i.e., if missing of `false` the button is not shown) so everything is working very nicely.

I do have some minor comments from which at least the command label one should be addressed. The others are more general thoughts and questions.

tsmaeder commented:
Hmh...I noticed that the shown actions are not correct when starting up with the view open in a view container with only one view. This is the case where the `ToolbarAwareTabBar` is showing the toolbar instead of the `ViewContainerPart`. The problem is that widget responsible for contributing to the toolbar is created lazily and we're not re-updating the toolbar at the right time in this case. I've tried to find a solution, but to no avail, so far.

tsmaeder commented:
@martin-fleck-at I addressed the comments and fixed the behavior when a single view is present in a view container: you can test this by opening the "Test View Drag and Drop" and then refreshing the browser.

martin-fleck-at reviewed (approved):
I re-tested the code and everything seems to be working now fine. Thank you @tsmaeder!
